common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,408 @@
+ *                    and the read proceeds by ensuring the entry is State.ReadInProgress
+ *      and agreeable for writes is Initial or Cached
+ *                    and the write proceeds by ensuring the entry is State.WriteInProgress
+ *      and agreeable for deletes is Cached
+ *
+ * 2. if entry is not in an agreeable state, then read- or write-around the cache;
+ *    for deletions, we allow the delete to proceed, and mark the entry as InvalidateWhenDone
+ *    the owning reader or writer is then responsible for ensuring that the entry is invalid
+ *    when that read or write completes
+ *
+ * 3. to swap in the new state to an existing entry, we use an AtomicReference.compareAndSet
+ *
+ * 4. only if the db operation completes with success, atomically set the state to Cached.
+ *
+ * 5. lastly, for cache invalidations that race with, we mark the entry as
+ *
+ */
+private object MultipleReadersSingleWriterCache {
+    /** Each entry has a state, as explained in the class comment above. */
+    object State extends Enumeration {
+        type State = Value
+        val ReadInProgress, WriteInProgress, InvalidateInProgress, InvalidateWhenDone, Cached = Value
+    }
+
+    import State._
+
+    /** Failure modes, which will only occur if there is a bug in this implementation */
+    case class ConcurrentOperationUnderRead(actualState: State) extends Exception(s"Cache read started, but completion raced with a concurrent operation: $actualState.")
+    case class ConcurrentOperationUnderUpdate(actualState: State) extends Exception(s"Cache update started, but completion raced with a concurrent operation: $actualState.")
+    case class SquashedInvalidation(actualState: State) extends Exception(s"Cache invalidation squashed due $actualState.")
+    case class StaleRead(actualState: State) extends Exception(s"Attempted read of invalid entry due to $actualState.")
+}
+
+trait MultipleReadersSingleWriterCache[W, Winfo] {
+    import MultipleReadersSingleWriterCache._
+    import MultipleReadersSingleWriterCache.State._
+
+    /** Subclasses: Toggle this to enable/disable caching for your entity type. */
+    protected lazy val cacheEnabled = true
+
+    /** Subclasses: tell me what key to use for updates. */
+    protected def cacheKeyForUpdate(w: W): Any
+
+    // Implicit conversions to boxed atomic references.
+    private implicit def toAtomic(state: State): AtomicReference[State] = new AtomicReference(state)
+    private implicit def toAtomic(value: Future[W]): AtomicReference[Future[W]] = new AtomicReference(value)
+    private implicit def toAtomic(transid: TransactionId): AtomicReference[TransactionId] = new AtomicReference(transid)
+
+    /**
+     * The entries in the cache will be a triple of (transid, State, Future[W]?).
+     *
+     * We need the transid in order to detect whether we have won the race to add an entry to the cache.
+     */
+    private class Entry(
```

Comment:
i think we were a bit lax here - the grab writer lock should've been synchronized. i removed the atomic refs in exchange for @volatile rather than synchronizing other methods as well.